central/auth/service/service_impl_test.go

```diff
@@ -527,21 +524,21 @@ func (s *authServiceAccessControlTestSuite) TestUpdateConfig() {
 		},
 	}
 
-	updatedConfig, err := s.svc.UpdateAuthMachineToMachineConfig(s.accessCtx,
+	_, err = s.svc.UpdateAuthMachineToMachineConfig(s.accessCtx,
 		&v1.UpdateAuthMachineToMachineConfigRequest{Config: config.GetConfig()})
 	s.NoError(err)
-	s.Equal(config.GetConfig(), updatedConfig.GetConfig())
+}
 
+func (s *authServiceAccessControlTestSuite) TestUpdateAddConfig() {
 	newConfig := &v1.AuthMachineToMachineConfig{
 		Id:                      "80c053c2-24a7-4b97-bd69-85b3a511241e",
```

Comment:
Here's something new. What's the difference between `svc.AddAuthMachineToMachineConfig` and `svc.UpdateAuthMachineToMachineConfig` when the latter one is provided with a config that did not previously exist?

Drumroll.

`AddAuthMachineToMachineConfig` ignores the `id` but `UpdateAuthMachineToMachineConfig` expects a non-empty `id`.
While you could argue this can be tested in `TestValidateAuthMachineToMachineConfig`, that would not be reliable. The difference is in the boolean flag that `AddAuthMachineToMachineConfig` and `UpdateAuthMachineToMachineConfig` provide to `validateAuthMachineToMachineConfig`. What happens when someone changes `validateAuthMachineToMachineConfig` to flip the argument meaning from `skipIDCheck` to `doIDCheck` and updates `AddAuthMachineToMachineConfig` but forgets about `UpdateAuthMachineToMachineConfig` (for example)?

`TestValidateAuthMachineToMachineConfig` tests will be corrected and will still pass. However, there are no `AddAuthMachineToMachineConfig` and `UpdateAuthMachineToMachineConfig` test that cover this behavior about the `id`.

Guess what's my thing here: I'd ask you to add tests covering this.

You may start feeling fatigue from creating repetitive tests and writing these `&v1.AuthMachineToMachineConfig{}` in particular. Well, it's in your power to change that (and you already started doing it): you can introduce `getSampleM2MConfig()` function for tests that creates a valid, fully populated config (if tests need, they can zero-out necessary attributes). You can have more specialized `getSampleGitHubM2MConfig()` and `getSampleGenericConfig()` for cases where such specialization makes sense. With this, many hand-written `&v1.AuthMachineToMachineConfig{...}` things could go away and test code shrink.

Why care about tests so much? It'll be good to talk about how it pays back, if you wonder :-)

Reply:
Added the test cases, I don't mind the repetition of instantiating a config and actually find it better this way over generic function + zeroing values for readability.